Clean up Solana indexing health on discovery

rickyrombo opened the pull request:
### Description

Normalize Solana indexer health checks to new pattern. Add aggregate_tips and payment_router to health check.

At the end of each successful job completion (even if no transactions were indexed), save the current timestamp to Redis. Then health_check ensures that the last successful job ran within a time diff. If indexing is stalled, whether due to celery not running the task, an error in indexing, or the current job being in an infinite loop, the timestamp won’t get updated and will slip the drift allowance.

This allows us to remove a ton of cruft:
- No longer need to fetch the global slot at the start of each indexing job
- No longer need to fetch the latest chain tx at the start of each indexing job separately from fetching the batch

...etc


---- 

- Adds payment_router and aggregate_tips to health check
- Normalize request args to "max_drift"
- Normalize indexer cache keys to be stored in a named tuple for better devex (easier autocomplete and type safety, and better organized/grouped)
- Nests all Solana indexer health checks under "solana_indexers" key
- Normalizes all Solana indexers to new format (**except plays**, which will be replaced soon by core plays anyway.  largely skipped because the indexer is different and we actually already use this one)
- Makes all Solana indexers affect `discovery_provider_healthy`
- Remove identity tip reactions indexing
- Remove implementation of SolanaNotifications (to remove dep on redis key so it could be cleaned up - left route and response structure intact and just return empty list for now)

### How Has This Been Tested?

Checked health check manually, ran tests

schottra commented on packages/discovery-provider/src/queries/health_check.py:
```diff
@@ -50,19 +50,21 @@ def health_check():
-        "reactions_max_last_reaction_drift": request.args.get(
-            "reactions_max_last_reaction_drift", type=int
+        "user_bank_max_drift": request.args.get(
+            "user_bank_max_drift", type=int, default=300
```
300 seconds seems...high? We're allowing a 5 minute dropoff on indexing before we consider it unhealthy?

rickyrombo replied:
It's def very high. Previously we weren't marking unhealthy by default _at all_ based on Solana indexers, so was being super relaxed to prevent noise. If we go five minutes without a succeed we know _for sure_ somethings completely busted and we should page/SDK should reselect etc. Healthz could set its own more narrow desired criteria via query params, like it does today with plays.

This could def be a conversation though - defaults are powerful and I'd be happy to make this like a single minute.

raymondjacobson replied:
I think as long as client isn't reselecting based off of this, it's probably ok to be 5 minutes

rickyrombo replied:
sorry can you say more? like do you mean if the client SDK reselects are based on specific query params not the default? or that client doesn't reselect based on solana indexing at all?

raymondjacobson replied:
I would worry that an issue on the solana side would just halt all of audius and prevent even reads from working on discovery - but these aren't even incorporated into the sdk selection atm, so it's a non-concern.

all the more reason to get the "acdc/core" layer to validate TX instead of doing it here

rickyrombo replied:
they are incorporated into SDK selection currently - the node gets marked as "behind" if the slot or time drifts are too far: https://github.com/AudiusProject/audius-protocol/blob/cc11317c5a3434bf77d4c218ca3463c4c9118368/packages/sdk/src/sdk/services/DiscoveryNodeSelector/healthChecks.ts#L57-L70

That doesn't preclude the SDK from selecting those nodes, it just tries to find a node that _isn't_ behind first.

The change here _does_ actually bubble up the unhealthy indexers to `discovery_provider_healthy` and to 500 - but I can revert that

raymondjacobson replied:
got it. i think that we will cause problems if we bubble that up atm, so i would say we should revert that

schottra reviewed (approved):
Generally a fan of the simplified approach.
1. I'm assuming you simulated error cases and tested that it actually reported unhealthy for all the services?
2. Does the 300 second setting match our existing health check criteria? i.e. if we had a max slot drift and a slot is approximately X seconds, does max_slot_drift * X ~= 300?

raymondjacobson reviewed (approved):
will we just set staging to have different values from prod?
thanks for doing this cleanup!

raymondjacobson commented:
> will we just set staging to have different values from prod? thanks for doing this cleanup!

nvm - the PR eliminates this. i get it now

rickyrombo commented:
- Removed defaults, so you have to opt in to the solana_indexers reporting "unhealthy" (which I kinda hate but I get it)
- Removed the 500 for when they do error
- Updated SDK health check types to look in the new place

Worth a small re-review for the new commits? @raymondjacobson / @schottra

schottra commented on packages/sdk/src/sdk/services/DiscoveryNodeSelector/healthChecks.test.ts:
```diff
@@ -150,7 +154,7 @@ describe('health_check', () => {
     })
 
-    // test('comms unhealthy', () => {
+    // it('comms unhealthy', () => {
```
This is curious.... changed from test->it but still commented? 😄

rickyrombo replied:
decided against re-enabling this in this pr, might do anohter though :)

raymondjacobson commented on packages/sdk/src/sdk/services/DiscoveryNodeSelector/healthCheckTypes.ts:
```diff
@@ -71,74 +71,61 @@ export type HealthCheckResponseData = DeepPartial<{
-          signature: string
-          slot: number
-          timestamp: number
+        slot_diff: number
```
i dont think we need this at all honestly

rickyrombo replied:
agree - but will wait to cut out the plays section until after core is converted anyway just to keep things easy

dharit-tan commented on packages/discovery-provider/src/tasks/index_aggregate_tips.py:
```diff
@@ -194,15 +192,15 @@ def index_rank_ups(
+        .order_by(UserTip.slot.desc())
+        .one()
+    )
+    (max_slot, last_tip_signature) = (
```
nit: i asked chatgpt and it said we could avoid the outer unpack:
`max_slot, last_tip_signature = max_slot_result if max_slot_result[0] is not None else (0, None)`

rickyrombo replied:
wdym outer unpack? these look identical to me save for the parens (which I think is autoformatted? could be wrong)

dharit-tan replied:
yeah it was the outer parens. nbd

dharit-tan reviewed (approved):
love all the deleted code
